src/runtime/hexagon_remote/halide_remote.cpp

```diff
@@ -45,15 +101,213 @@ int halide_do_task(void *user_context, halide_task_t f, int idx,
+            // and tries to do it. If it cannot find a job, it goes
+            // to sleep.
+            job = work_queue.jobs;
+            if (!job) {
+                // release the lock and go to sleep.
+                // ***************************
+                // *** Work queue unlocked ***
+                // ***************************
+                FARF(LOW, "HVX_TP: %d: goto_work: Couldn't find a job, going to sleep\n", tid);
+                work_queue.unlock();
+                qurt_sem_down(&wait_for_work);
+                FARF(LOW, "HVX_TP: %d: goto_work: Got woken up. Going to look for work.\n", tid);
+                 // ***********************
+                // *** Lock work queue ***
+                // ***********************
+                work_queue.lock();
+                continue;
+            } else {
+                FARF(LOW, "HVX_TP: %d: goto_work: Found a job working on x = %d\n", tid, job->next);
+            }
+        } else {
+            // We are here only if this thread owns a job that is not done.
+            // So do a part of it.
+            if (owned_job->claimed()) {
+                // the owner goes to sleep till the workers wake it up.
+                FARF(LOW, "HVX_TP: %d: goto_work: Owner about to sleep\n", tid);
+                qurt_cond_wait(&owned_job->wakeup_owner, &work_queue.work_mutex);
+                FARF(LOW, "HVX_TP: %d: goto_work: Owner waking up.\n", tid);
+                // This thread should have the lock now after having been woken up.
+                break;
+            } else {
+                job = owned_job;
+                FARF(LOW, "HVX_TP: %d: goto_work: Owner about to work\n", tid);
+            }
+        }
+        int myjob = job->next++;
+        job->active_workers++;
+
+        // If all tasks of the job have been claimed, then pop the job off the stack.
+        if (job->claimed()) {
+            work_queue.jobs = job->next_job;
+            FARF(LOW, "HVX_TP: %d: goto_work: All tasks in job claimed. Popping job from stack.\n", tid);
+        }
+        // ***************************
+        // *** Work queue unlocked ***
+        // ***************************
+        work_queue.unlock();
+
+        FARF(LOW, "HVX_TP: %d: goto_work: About to do_task, user_context = 0x%x, f = 0x%x x = %d \n", tid, job->user_context, job->f, myjob);
+        int result = halide_do_task(job->user_context, job->f,
+                                    myjob, job->closure);
+        FARF(LOW, "HVX_TP: %d: goto_work: Finished do_task with status = %d\n", tid, result);
+        // ***********************
+        // *** Lock work queue ***
+        // ***********************
+        work_queue.lock();
         if (result) {
-            return result;
+            job->exit_status = result;
+            job->next = job->end;
+            work_queue.jobs = job->next_job;
+            work_queue.shutdown = true;
```

Comment:
That will turn off the thread pool. If we're just cancelling work, that shouldn't happen. I think it would be simpler to just have it skip any remaining work, return the exit code, but not change anything else.